core/src/mender-client.c

```diff
@@ -926,51 +974,72 @@ mender_compare_device_types(const char  *device_type_artifact,
     mender_log_error("None of the device types from the deployment are compatible with device '%s'", device_type_device);
     return MENDER_FAIL;
 }
-#endif
 
-#ifdef CONFIG_MENDER_FULL_PARSE_ARTIFACT
 #ifdef CONFIG_MENDER_PROVIDES_DEPENDS
 static mender_err_t
-mender_store_provides(mender_artifact_ctx_t *mender_artifact_ctx) {
+mender_filter_provides(mender_artifact_ctx_t *mender_artifact_ctx, mender_key_value_list_t **new_provides, mender_key_value_list_t **stored_provides) {
 
-    assert(NULL != mender_artifact_ctx);
     mender_err_t ret = MENDER_FAIL;
+    /* Clears provides */
+    bool matches;
+    for (size_t i = 0; i < mender_artifact_ctx->payloads.size; i++) {
+        for (size_t j = 0; j < mender_artifact_ctx->payloads.values[i].clears_provides_size; j++) {
+            const char *to_clear = mender_artifact_ctx->payloads.values[i].clears_provides[j];
+            for (mender_key_value_list_t *item = *stored_provides; NULL != item; item = item->next) {
+                if (MENDER_OK != mender_utils_compare_wildcard(item->key, to_clear, &matches)) {
+                    mender_log_error("Unable to compare wildcard");
```

Comment:
The error message should mention what kind of wildcard it was and what it was doing. Otherwise it's quite useless for the user.